ops/testing.py

```diff
@@ -834,5 +857,332 @@ def network_get(self, endpoint_name, relation_id=None):
+            client = _TestingPebbleClient(self)
+            self._pebble_clients[socket_path] = client
+        return client
+
+
+class _TestingPebbleClient:
+    """This conforms to the interface for pebble.Client but provides canned data.
+
+    DO NOT use this class directly, it is used by `Harness`_ to run interactions with Pebble.
+    `Harness`_ is responsible for maintaining the internal consistency of the values here,
+    as the only public methods of this type are for implementing Client.
+    """
+
+    def __init__(self, backend: _TestingModelBackend):
+        self._backend = _TestingModelBackend
+        self._layers = {}
+        # Has a service been started/stopped?
+        self._service_status = {}
+
+    def get_system_info(self) -> pebble.SystemInfo:
+        """Get system info."""
+        raise NotImplementedError(self.get_system_info)
+
+    def get_warnings(
+            self, select: pebble.WarningState = pebble.WarningState.PENDING,
+    ) -> typing.List['pebble.Warning']:
+        """Get list of warnings in given state (pending or all)."""
+        raise NotImplementedError(self.get_warnings)
+        query = {'select': select.value}
+        result = self._request('GET', '/v1/warnings', query)
+        return [Warning.from_dict(w) for w in result['result']]
+
+    def ack_warnings(self, timestamp: datetime.datetime) -> int:
+        """Acknowledge warnings up to given timestamp, return number acknowledged."""
+        raise NotImplementedError(self.ack_warnings)
+        body = {'action': 'okay', 'timestamp': timestamp.isoformat()}
+        result = self._request('POST', '/v1/warnings', body=body)
+        return result['result']
+
+    def get_changes(
+            self, select: pebble.ChangeState = pebble.ChangeState.IN_PROGRESS, service: str = None,
+    ) -> typing.List[pebble.Change]:
+        """Get list of changes in given state, filter by service name if given."""
+        raise NotImplementedError(self.get_changes)
+
+    def get_change(self, change_id: pebble.ChangeID) -> pebble.Change:
+        """Get single change by ID."""
+        raise NotImplementedError(self.get_changes)
+
+    def abort_change(self, change_id: pebble.ChangeID) -> pebble.Change:
+        """Abort change with given ID."""
+        raise NotImplementedError(self.abort_change)
+
+    def autostart_services(self, timeout: float = 30.0, delay: float = 0.1) -> pebble.ChangeID:
+        """Start the startup-enabled services and wait (poll) for them to be started.
+
+        Raises ChangeError if one or more of the services didn't start. If
+        timeout is 0, submit the action but don't wait; just return the change
+        ID immediately.
+        """
+        for name, service in self._render_services().items():
+            # TODO: jam 2021-04-20 This feels awkward that Service.startup might be a string or
+            #  might be an enum. Probably should make Service.startup a property rather than an
+            #  attribute.
```

Comment:
Hmm, I hadn't considered that startup might be set to an enum (I believe I added the ServiceStartup enum later). However, I think what we should do is make it always an enum -- and on `__init__` it should be parsed from string into the enum, and on `to_dict` the reverse.

Reply:
I'm going to leave it for now, but it is something we can do quickly as a follow up.